### PR TITLE
t/335: Created the Content Security Policy guide

### LINKED ISSUE #335: Support CSP
CSP has been a serious problem for CKEditor 4 (https://dev.ckeditor.com/ticket/8584) due to some old decisions that's been made (mainly due to legacy browsers support).

In case of CKEditor 5 I don't expect to see any problems, but we should test for it. AFAICS, it should be possible to simply enforce some CSP settings using a `<meta>` tag:

``` html
<meta http-equiv="Content-Security-Policy" content="default-src https://cdn.example.net; child-src 'none'; object-src 'none'">
```
- https://github.com/apache/cordova-plugin-whitelist#content-security-policy
- https://bugzilla.mozilla.org/show_bug.cgi?id=663570

If not, then let's create a simple test env (a Node.js HTTP server and a sample).

### PR DESCRIPTION
**Pull request by oleq**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Docs: Created the Content Security Policy guide. Closes #335.

---

### Additional information

* https://github.com/ckeditor/ckeditor5-dev/pull/482 enabled CSP in all manual tests
* https://github.com/ckeditor/ckeditor5-ckfinder/pull/36 changes in tests
* https://github.com/ckeditor/ckeditor5-theme-lark/pull/218 changes in tests

### CODE REVIEW

File: docs/builds/guides/integration/csp.md [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+* `style-src 'self' 'unsafe-inline'`: `'unsafe-inline'` is necessary for:
+	* webpack's [style-loader](https://github.com/webpack-contrib/style-loader) to load {@link framework/guides/theme-customization#styles-processing-and-bundling editor UI styles}.
+
+    **Note**: You can {@link builds/guides/integration/advanced-setup#option-extracting-css extract styles to a separate `.css` file} during the editor building process and remove this directive.
+	* certain editor content styles to work properly. For instance, you are going to need it if you want to enable editor features like {@link features/font Font} or {@link features/text-alignment Text alignment} or any other feature that uses inline `style="..."` attributes in the content.
+* `frame-src *`: Necessary for the {@link features/media-embed Media embed} feature to load media with previews (containing `<iframe>`).
+
+    **Note**: Use the more strict `frame-src 'self'` if all the media in the edited content come from the same domain as your application.
+
+## Minimal configuration
```
Comment (Reinmar) [inline docs/builds/guides/integration/csp.md]: Minimal sounds like "the weakest". I'd try something like "Strictest possible configuration".

File: docs/builds/guides/integration/csp.md [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+## Impact of CSP on editor features
+
+Some CSP directives have an impact on certain rich text editor features. Here's the round-up of directives and their specific roles in the editor:
+
+* `default-src 'none'`: resets the policy and blocks everything. All successive directives work as a white–list. By itself, as long as followed by other directives, it has no impact on the editor.
+* `connect-src 'self'`
+	* Allows the {@link features/image-upload editor upload features} to use [XMLHttpReqests](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) (AJAX) to upload files to the server, for instance when an image is pasted or dropped into the editor content. The `'self`' value ensures the requests remain within the same host.
+	* Allows {@link builds/guides/integration/saving-data#autosave-feature auto–saving editor data} using XMLHttpReqests.
+
+	**Note**: To use {@link features/ckfinder CKFinder} hosted from another domain, make sure the domain is also included in the directive value.
```
Comment (Reinmar) [inline docs/builds/guides/integration/csp.md]: Extract all things about CKF to its own guide. It's not CKE's issue. Link to that guide, though.

File: docs/builds/guides/integration/csp.md [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+
+    **Note**: You can {@link builds/guides/integration/advanced-setup#option-extracting-css extract styles to a separate `.css` file} during the editor building process and remove this directive.
+	* certain editor content styles to work properly. For instance, you are going to need it if you want to enable editor features like {@link features/font Font} or {@link features/text-alignment Text alignment} or any other feature that uses inline `style="..."` attributes in the content.
+* `frame-src *`: Necessary for the {@link features/media-embed Media embed} feature to load media with previews (containing `<iframe>`).
+
+    **Note**: Use the more strict `frame-src 'self'` if all the media in the edited content come from the same domain as your application.
+
+## Minimal configuration
+
+Knowing the role of each directive, the most strict set of rules that allows CKEditor 5 to run is as follows:
```
Comment (Reinmar) [inline docs/builds/guides/integration/csp.md]: s/most strict/strictest

File: docs/builds/guides/integration/csp.md [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+The recommended CSP configuration that allows the rich text editor to run out–of–the–box with all standard features using the content like images or media from external hosts looks as follows:
+
+```
+default-src 'none'; connect-src 'self'; script-src 'self'; img-src * data:; style-src 'self' 'unsafe-inline'; frame-src *
+```
+
+## Impact of CSP on editor features
+
+Some CSP directives have an impact on certain rich text editor features. Here's the round-up of directives and their specific roles in the editor:
```
Comment (Reinmar) [inline docs/builds/guides/integration/csp.md]:
```suggestion
Some CSP directives have an impact on certain rich text editor features. Here is the round-up of directives and their specific roles in the editor:
```

File: docs/builds/guides/integration/csp.md [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+* `default-src 'none'`: resets the policy and blocks everything. All successive directives work as a white–list. By itself, as long as followed by other directives, it has no impact on the editor.
+* `connect-src 'self'`
+	* Allows the {@link features/image-upload editor upload features} to use [XMLHttpReqests](https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest) (AJAX) to upload files to the server, for instance when an image is pasted or dropped into the editor content. The `'self`' value ensures the requests remain within the same host.
+	* Allows {@link builds/guides/integration/saving-data#autosave-feature auto–saving editor data} using XMLHttpReqests.
```
Comment (Reinmar) [inline docs/builds/guides/integration/csp.md]:
```suggestion
	* Allows {@link builds/guides/integration/saving-data#autosave-feature auto–saving editor data} using `XMLHttpRequest`.
```

File: docs/builds/guides/integration/csp.md (hunk 2 above)
Comment (oleq) [reply docs/builds/guides/integration/csp.md]: I removed the notes about CKF. To be linked when an appropriate guide in CKF is created.

**Comment by wwalc** [comment]
I'm unable to review it, feel free to merge it if it looks okay to you.

> A code to be added when the corresponding section lands in CKF docs
> 
> ```
> <info-box>
> 	A different set of Content Security Policy directives might be necessary to run {@link features/ckfinder CKFinder} along with CKEditor 5. Check out the file manager {@link @ckfinder ckfinder3/#!/guide/dev_integration-csp-directives-required-by-ckfinder documentation} to learn more.
> </info-box>
> ```

It looks like the URL is a bit broken, the correct one is https://ckeditor.com/docs/ckfinder/ckfinder3/#!/guide/dev_integration-section-csp-directives-required-by-ckfinder

(there is "-section" in the middle)

**Pull request merged**